File: test/jvm/com/cattailsw/nanidroid/install/NarFilesystemInspectorTest.java
package com.cattailsw.nanidroid.install;

import static org.junit.Assert.*;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.Test;

public final class NarFilesystemInspectorTest {
    private static NarFilesystemInspector.Result result(
            int state, int error, String[] paths, int[] types, long[] facts) {
        return NarFilesystemInspector.fromNative(
                state, error, 0, paths.length, 7, paths, types, facts);
    }

    private static NarFilesystemInspector inspector(
            NarFilesystemInspector.Loader loader,
            NarFilesystemInspector.Backend backend) {
        return new NarFilesystemInspector(loader, backend);
    }

    @Test
    public void backendIsLazyOrderedImmutableAndDefensivelyCopied() {
        AtomicInteger loads = new AtomicInteger();
        AtomicInteger calls = new AtomicInteger();
        String[] paths = {"empty", "nested/file-\ud83d\ude00"};
        int[] types = {2, 1};
        long[] facts = {0, 11, 12, 7, 21, 22};
        NarFilesystemInspector.Result made = result(2, 0, paths, types, facts);
        NarFilesystemInspector value = inspector(
                loads::incrementAndGet,
                (root, target) -> { calls.incrementAndGet(); return made; });
        paths[0] = "changed";
        types[0] = 1;
        facts[0] = 99;

        NarFilesystemInspector.TrustedRoot root =
                new NarFilesystemInspector.TrustedRoot("/trusted/root");
        NarFilesystemInspector.Result first = value.inspect(root, "ghost");
        assertSame(first, value.inspect(root, "ghost"));
        assertEquals(1, loads.get());
        assertEquals(2, calls.get());
        assertEquals(NarFilesystemInspector.State.PRESENT, first.state());
        assertEquals(NarFilesystemInspector.Error.OK, first.error());
        assertEquals(2, first.entryCount());
        assertEquals(7, first.totalFileSize());
        List<NarFilesystemInspector.Entry> entries = first.entries();
        assertEquals("empty", entries.get(0).path());
        assertEquals(NarFilesystemInspector.Type.DIRECTORY, entries.get(0).type());
        assertEquals(0, entries.get(0).size());
        assertEquals(11, entries.get(0).device());
        assertEquals(12, entries.get(0).inode());
        assertEquals("nested/file-\ud83d\ude00", entries.get(1).path());
        assertThrows(UnsupportedOperationException.class, () -> entries.clear());
    }

    @Test
    public void loaderAndRuntimeFailuresAreTypedButOomePropagates() {
        NarFilesystemInspector.TrustedRoot root =
                new NarFilesystemInspector.TrustedRoot("/trusted");
        NarFilesystemInspector.Result link = inspector(
                () -> { throw new UnsatisfiedLinkError("missing"); },
                (r, t) -> { throw new AssertionError(); }).inspect(root, "x");
        assertEquals(NarFilesystemInspector.Error.LINKAGE, link.error());
        NarFilesystemInspector.Result security = inspector(
                () -> { throw new SecurityException("denied"); },
                (r, t) -> { throw new AssertionError(); }).inspect(root, "x");
        assertEquals(NarFilesystemInspector.Error.SECURITY, security.error());
        NarFilesystemInspector.Result nativeFailure = inspector(
                () -> {}, (r, t) -> null).inspect(root, "x");
        assertEquals(NarFilesystemInspector.Error.NATIVE, nativeFailure.error());
        assertThrows(OutOfMemoryError.class, () -> inspector(
                () -> { throw new OutOfMemoryError(); },
                (r, t) -> null).inspect(root, "x"));
    }

    @Test
    public void nativeCodesAndMalformedDtosHaveStableTypedResults() {
        assertEquals(NarFilesystemInspector.State.ABSENT,
                result(1, 0, new String[0], new int[0], new long[0]).state());
        for (int code = 0; code <= 21; code++) {
            assertNotNull(result(0, code, new String[0], new int[0], new long[0]).error());
        }
        NarFilesystemInspector.Result malformed = NarFilesystemInspector.fromNative(
                2, 0, 0, 1, 0, new String[]{"x"}, new int[0], new long[0]);
        assertEquals(NarFilesystemInspector.State.ERROR, malformed.state());
        assertEquals(NarFilesystemInspector.Error.NATIVE, malformed.error());
        assertTrue(malformed.entries().isEmpty());
    }

    @Test
    public void packageSeamExposesNoPublicCapabilityOrResourceTypes() {
        assertFalse(Modifier.isPublic(NarFilesystemInspector.class.getModifiers()));
        for (Constructor<?> constructor : NarFilesystemInspector.class.getDeclaredConstructors()) {
            assertFalse(Modifier.isPublic(constructor.getModifiers()));
        }
        for (Class<?> nested : NarFilesystemInspector.class.getDeclaredClasses()) {
            assertFalse(Modifier.isPublic(nested.getModifiers()));
        }
        for (Method method : NarFilesystemInspector.class.getDeclaredMethods()) {
            assertFalse(Modifier.isPublic(method.getModifiers()));
            assertFalse(method.getReturnType().getName().matches(
                    ".*(File|Descriptor|Stream|Channel|Pointer).*"));
        }
        for (Field field : NarFilesystemInspector.Entry.class.getDeclaredFields()) {
            assertTrue(Modifier.isPrivate(field.getModifiers()));
            assertFalse(field.getType().getName().matches(
                    ".*(File|Descriptor|Stream|Channel|Pointer).*"));
        }
    }
}

File: src/test/java/com/cattailsw/nanidroid/install/NarFilesystemInspectorTest.kt
package com.cattailsw.nanidroid.install

import org.junit.Assert
import org.junit.Test
import org.junit.function.ThrowingRunnable
import java.lang.reflect.Modifier
import java.util.concurrent.atomic.AtomicInteger

class NarFilesystemInspectorTest {
    @Test
    fun backendIsLazyOrderedImmutableAndDefensivelyCopied() {
        val loads = AtomicInteger()
        val calls = AtomicInteger()
        val paths = arrayOf<String>("empty", "nested/file-\ud83d\ude00")
        val types = intArrayOf(2, 1)
        val facts = longArrayOf(0, 11, 12, 7, 21, 22)
        val made: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result =
            result(2, 0, paths, types, facts)
        val value: com.cattailsw.nanidroid.install.NarFilesystemInspector = inspector(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Loader { loads.incrementAndGet() },
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Backend { root: String?, target: String? ->
                calls.incrementAndGet()
                made
            })
        paths[0] = "changed"
        types[0] = 1
        facts[0] = 99

        val root: com.cattailsw.nanidroid.install.NarFilesystemInspector.TrustedRoot =
            com.cattailsw.nanidroid.install.NarFilesystemInspector.TrustedRoot("/trusted/root")
        val first: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result =
            value.inspect(root, "ghost")
        Assert.assertSame(first, value.inspect(root, "ghost"))
        Assert.assertEquals(1, loads.get().toLong())
        Assert.assertEquals(2, calls.get().toLong())
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.State.PRESENT,
            first.state()
        )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.OK,
            first.error()
        )
        Assert.assertEquals(2, first.entryCount().toLong())
        Assert.assertEquals(7, first.totalFileSize())
        val entries: List<com.cattailsw.nanidroid.install.NarFilesystemInspector.Entry> =
            first.entries()
        Assert.assertEquals("empty", entries.get(0).path())
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Type.DIRECTORY,
            entries.get(0).type()
        )
        Assert.assertEquals(0, entries.get(0).size())
        Assert.assertEquals(11, entries.get(0).device())
        Assert.assertEquals(12, entries.get(0).inode())
        Assert.assertEquals("nested/file-\ud83d\ude00", entries.get(1).path())
        Assert.assertThrows(
            UnsupportedOperationException::class.java,
            ThrowingRunnable { (entries as MutableList<com.cattailsw.nanidroid.install.NarFilesystemInspector.Entry>).clear() })
    }

    @Test
    fun loaderAndRuntimeFailuresAreTypedButOomePropagates() {
        val root: com.cattailsw.nanidroid.install.NarFilesystemInspector.TrustedRoot =
            com.cattailsw.nanidroid.install.NarFilesystemInspector.TrustedRoot("/trusted")
        val link: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result = inspector(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Loader {
                throw UnsatisfiedLinkError(
                    "missing"
                )
            },
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Backend { r: String?, t: String? -> throw AssertionError() }).inspect(
            root,
            "x"
        )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.LINKAGE,
            link.error()
        )
        val security: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result = inspector(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Loader {
                throw SecurityException(
                    "denied"
                )
            },
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Backend { r: String?, t: String? -> throw AssertionError() }).inspect(
            root,
            "x"
        )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.SECURITY,
            security.error()
        )
        val nativeFailure: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result =
            inspector(
                com.cattailsw.nanidroid.install.NarFilesystemInspector.Loader {},
                com.cattailsw.nanidroid.install.NarFilesystemInspector.Backend { r: String?, t: String? -> null }).inspect(
                root,
                "x"
            )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.NATIVE,
            nativeFailure.error()
        )
        Assert.assertThrows(OutOfMemoryError::class.java, ThrowingRunnable {
            inspector(
                com.cattailsw.nanidroid.install.NarFilesystemInspector.Loader { throw OutOfMemoryError() },
                com.cattailsw.nanidroid.install.NarFilesystemInspector.Backend { r: String?, t: String? -> null }).inspect(
                root,
                "x"
            )
        })
    }

    @Test
    fun nativeCodesAndMalformedDtosHaveStableTypedResults() {
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.State.ABSENT,
            result(1, 0, emptyArray<String>(), IntArray(0), LongArray(0)).state()
        )
        for (code in 0..20) {
            Assert.assertNotNull(
                result(
                    0,
                    code,
                    emptyArray<String>(),
                    IntArray(0),
                    LongArray(0)
                ).error()
            )
        }
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.INPUT,
            result(0, 100, emptyArray<String>(), IntArray(0), LongArray(0)).error()
        )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.SECURITY,
            result(0, 103, emptyArray<String>(), IntArray(0), LongArray(0)).error()
        )
        val malformed: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result =
            com.cattailsw.nanidroid.install.NarFilesystemInspector.fromNative(
                2, 0, 0, 1, 0, arrayOf<String>("x"), IntArray(0), LongArray(0)
            )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.State.ERROR,
            malformed.state()
        )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.NATIVE,
            malformed.error()
        )
        Assert.assertTrue(malformed.entries().isEmpty())
        val nullPath: com.cattailsw.nanidroid.install.NarFilesystemInspector.Result =
            com.cattailsw.nanidroid.install.NarFilesystemInspector.fromNative(
                2, 0, 0, 1, 0, (arrayOfNulls<String>(1) as Array<String>), intArrayOf(1),
                longArrayOf(0, 0, 0)
            )
        Assert.assertEquals(
            com.cattailsw.nanidroid.install.NarFilesystemInspector.Error.NATIVE,
            nullPath.error()
        )
    }

    @Test
    fun kotlinPackageSeamExposesNoFilesystemCapabilityTypes() {
        Assert.assertNotNull(
            com.cattailsw.nanidroid.install.NarFilesystemInspector::class.java.getAnnotation<Metadata?>(
                Metadata::class.java
            )
        )
        for (nested in com.cattailsw.nanidroid.install.NarFilesystemInspector::class.java.getDeclaredClasses()) {
            Assert.assertNotNull(nested.getAnnotation<Metadata?>(Metadata::class.java))
        }
        for (method in com.cattailsw.nanidroid.install.NarFilesystemInspector::class.java.getDeclaredMethods()) {
            Assert.assertFalse(
                method.getReturnType().getName()
                    .matches("(java\\.io\\..*|java\\.nio\\.channels\\..*)".toRegex())
            )
        }
        for (field in com.cattailsw.nanidroid.install.NarFilesystemInspector.Entry::class.java.getDeclaredFields()) {
            if (field.isSynthetic() || field.getName() == "\$stable") continue
            Assert.assertTrue(Modifier.isPrivate(field.getModifiers()))
            Assert.assertFalse(
                field.getType().getName()
                    .matches("(java\\.io\\..*|java\\.nio\\.channels\\..*)".toRegex())
            )
        }
    }

    companion object {
        private fun result(
            state: Int, error: Int, paths: Array<String>, types: IntArray?, facts: LongArray?
        ): com.cattailsw.nanidroid.install.NarFilesystemInspector.Result {
            return com.cattailsw.nanidroid.install.NarFilesystemInspector.fromNative(
                state, error, 0, paths.size, 7, paths, types, facts
            )
        }

        private fun inspector(
            loader: com.cattailsw.nanidroid.install.NarFilesystemInspector.Loader,
            backend: com.cattailsw.nanidroid.install.NarFilesystemInspector.Backend
        ): com.cattailsw.nanidroid.install.NarFilesystemInspector {
            return com.cattailsw.nanidroid.install.NarFilesystemInspector(loader, backend)
        }
    }
}
